### PR TITLE
fix(utils): fix resources map becomes directory

### LINKED ISSUE #10187: [bug]: resources map unintuitively becomes directory
### Describe the bug

The following config:

```json
{
  "bundle": {
    "resources": {
      "../resources/user.json": "resources/user.json"
    }
  }
}
```

Results in:

```
src-tauri/
├─ target/
│  ├─ debug/
│  │  ├─ resources/
│  │  │  ├─ user.json/
│  │  │  │  ├─ user.json
```

### Reproduction

```bash
$ cargo create-tauri-app --beta
  my-app
$ cd my-app
$ mkdir resources
$ touch resources/user.json
$ nano src-tauri/tauri.conf.json
  {
    "productName": "my-app",
    "version": "0.0.0",
    "identifier": "com.my-app.www",
    "build": {
      "beforeDevCommand": "npm run dev",
      "devUrl": "http://localhost:1420",
      "beforeBuildCommand": "npm run build",
      "frontendDist": "../dist"
    },
    "app": {
      "windows": [
        {
          "title": "superboard",
          "width": 800,
          "height": 600
        }
      ],
      "security": {
        "csp": null
      }
    },
    "bundle": {
      "active": true,
      "targets": "all",
      "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ],
      "resources": {
        "../resources/user.json": "resources/user.json"
      }
    }
  }
$ cargo tauri dev
```

### Expected behavior

```
src-tauri/
├─ target/
│  ├─ debug/
│  │  ├─ resources/
│  │  │  ├─ user.json
```

### Full `tauri info` output

```text
[✔] Environment
    - OS: Ubuntu 22.4.0 X64
    ✔ webkit2gtk-4.1: 2.44.2
    ✔ rsvg2: 2.52.5
    ✔ rustc: 1.78.0 (9b00956e5 2024-04-29)
    ✔ cargo: 1.78.0 (54d8815d0 2024-03-26)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-unknown-linux-gnu (environment override by RUSTUP_TOOLCHAIN)
    - node: 22.2.0
    - pnpm: 9.1.1
    - npm: 10.7.0
    - bun: 1.1.0

[-] Packages
    - tauri [RUST]: 2.0.0-beta.22
    - tauri-build [RUST]: 2.0.0-beta.17
    - wry [RUST]: 0.40.1
    - tao [RUST]: 0.28.1
    - tauri-cli [RUST]: 2.0.0-beta.14
    - @tauri-apps/api [NPM]: 2.0.0-beta.13
    - @tauri-apps/cli [NPM]: 2.0.0-beta.14

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../dist
    - devUrl: http://localhost:1420/
    - framework: React
    - bundler: Vite
```

### Stack trace

_No response_

### Additional context

If this **is** the expected behaviour, then these docs are misleading: https://v2.tauri.app/develop/resources/

### PR DESCRIPTION
**Pull request by amrbashir**
closes #10187

Fixes the behavior of mapped resources generating extra directory, for example: `"../resources/user.json": "resources/user.json"` generates this resource `resources/user.json/user.json` where it should generate `resources/user.json`

This PR includes a refactor of the Iterator implementation which splits it into more scoped functions and relis on recursing instead of a loop which makes the code a lot more readable and easier to maintain.

### CODE REVIEW

**Comment by pewsheen** [comment]
I tested with some folder cases, and it seems to have some differences from the past version when the source has folders inside a folder or the destination is a folder:

```
src/
├─ assets/
│  ├─ lang/
│  │  ├─ de.json
│  ├─ tauri.svg
```

### Source: Folder, Dest: Folder

```json
{
  "resources": {
    "assets/": "dist/"
  }  
}
```

```
# before
Resource { path: "assets/tauri.svg", target: "dist/tauri.svg" }
Resource { path: "assets/lang/de.json", target: "dist/lang/de.json" }
# now
Resource { path: "assets/tauri.svg", target: "dist/" }
Resource { path: "assets/lang/de.json", target: "dist/" }
```

### Source: Folder, Dest: Root

```json
{
  "resources": {
    "assets/": ""
  }  
}
```

```
# before
Resource { path: "assets/tauri.svg", target: "tauri.svg" }
Resource { path: "assets/lang/de.json", target: "lang/de.json" }
# now
Resource { path: "assets/tauri.svg", target: "tauri.svg" }
Resource { path: "assets/lang/de.json", target: "de.json" }
```

### Source: Glob, Dest: Folder

```json
{
  "resources": {
    "assets/*": "dist/"
  }  
}
```

```
# before
Resource { path: "assets/tauri.svg", target: "dist/tauri.svg" }
Resource { path: "assets/lang/de.json", target: "dist/lang/de.json" }
# now
Resource { path: "assets/tauri.svg", target: "dist/" }
Resource { path: "assets/lang/de.json", target: "dist/" }
```

### Source: Glob, Dest: Root (this one is correct)

```json
{
  "resources": {
    "assets/*": ""
  }  
}
```

```
# before
Resource { path: "assets/tauri.svg", target: "tauri.svg" }
Resource { path: "assets/lang/de.json", target: "de.json" }
# now
Resource { path: "assets/tauri.svg", target: "tauri.svg" }
Resource { path: "assets/lang/de.json", target: "de.json" }
```

**Comment by amrbashir** [comment]
Thanks @pewsheen for extensive tests, I will fix the `Source: Folder, Dest: Folder or Root` cases.

However `Source: Glob, Dest: Folder` is basically the same as `Source: Glob, Dest: Root` and so they should act the same, and so I consider `Source: Glob, Dest: Folder` got fixed from previous behavior.

**Comment by amrbashir** [comment]
I just noticed that `Source: Glob, Dest: Folder` is just plain broken, it should be 
```
# before
Resource { path: "assets/tauri.svg", target: "dist/tauri.svg" }
Resource { path: "assets/lang/de.json", target: "dist/lang/de.json" }
# now
Resource { path: "assets/tauri.svg", target: "dist/tauri.svg" }
Resource { path: "assets/lang/de.json", target: "dist/de.json" }
```
which is a different behavior than before but consistent with `Source Glob, Dest: Root`

**Comment by amrbashir** [comment]
@pewsheen all should be good now, please give it another test

**Comment by pewsheen** [comment]
This one is weird

```
("../src/sounds/*".into(), "voices".into()),
```

```
Resource {
    path: "../src/sounds/lang/fr.wav",
    target: "voices/../src/sounds/lang/fr.wav",
},
Resource {
    path: "../src/sounds/lang/es.wav",
    target: "voices/../src/sounds/lang/es.wav",
},
```

And if the the glob path is not valid, it'll remain in the glob format

```
Resource {
    path: "not/exists/path/*",
    target: "voices",
},
```

Maybe we should check the glob path is valid or return `crate::Error::GlobPathNotFound`

**Comment by amrbashir** [comment]
@pewsheen these two cases should be fixed now, let me know if you find any more bugs

**Review by pewsheen — APPROVED** [review]
Looks good!

File: .changes/resources-map-becoming-dirs.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"tauri-utils": "patch:bug"
+---
+
+Fix `ResourcePaths` iterator returning an unexptected result for mapped resources, for example `"../resources/user.json": "resources/user.json"` generates this resource `resources/user.json/user.json` where it should generate just `resources/user.json`.
```
Comment (lucasfernog) [inline .changes/resources-map-becoming-dirs.md]:
```suggestion
Fix `ResourcePaths` iterator returning an unexpected result for mapped resources, for example `"../resources/user.json": "resources/user.json"` generates this resource `resources/user.json/user.json` where it should generate just `resources/user.json`.
```

**Review by lucasfernog — APPROVED** [review]
I opened this PR considering writing tests to validate this, but you already did it ❤️

**Comment by Raicuparta** [comment]
I think this PR is probably what broke the resources setting for me. I have a `resources` folder that has several other folders inside it, and those folders have a bunch of files at different levels. But the root `resources` folder doesn't have any files itself. My tauri config had `"resources": ["resources/*"]`, and that stopped working, with the error `resource path resources\* doesn't exist`. I had to change my setting to `"resources": ["resources/**/*"]` to get it working again.

I should probably make a separate issue for it, but don't have the time to make it detailed enough. And sorry if this is unrelated (very possible)

**Comment by FabianLars** [comment]
@Raicuparta 

> I should probably make a separate issue for it, but don't have the time to make it detailed enough

What you wrote here is already detailed enough for an initial issue. :)

**Comment by amrbashir** [comment]
@Raicuparta I believe `"resources": ["resources/**/*"]` should be the correct behavior, and "resources": ["resources/*"] should only match files directly under `resources` dir. This is how globbing behaves in bash as well.

However I think we may need to improve the error.

**Comment by Raicuparta** [comment]
That makes sense. I guess the error could specifically mention globs when those are used to make it more clear. But maybe the breaking change could be documented in V2 migration docs. Unless it's considered a bug fix because the previous version wasn't following globbing expectations.

**Comment by amrbashir** [comment]
I will open a PR to improve the error a bit, However for `./resources/*` it will no longer print any errors since the glob is correct but it just couldn't find any files there which is the correct behavior here. 

The migration docs will be updated.

**Pull request merged**